src/control/lib/helper/app.go

```diff
@@ -0,0 +1,246 @@
+	a.configureLogging(path)
+	return a
+}
+
+// configureLogging configures the application's logger. If logPath
+// is empty, non-error messages are not logged.
+func (a *App) configureLogging(logPath string) {
+	logLevel := logging.LogLevelError
+	combinedOut := ioutil.Discard
+	if logPath != "" {
+		lf, err := common.AppendFile(logPath)
+		if err == nil {
+			combinedOut = lf
+			logLevel = logging.LogLevelDebug
+		}
+	}
+
+	// By default, we only want to log errors to stderr.
+	a.log = logging.NewCombinedLogger(a.Name(), combinedOut).
+		WithErrorLogger(logging.NewCommandLineErrorLogger(os.Stderr)).
+		WithLogLevel(logLevel)
+}
+
+// WithAllowedCallers adds a list of process names allowed to call this
+// application.
+func (a *App) WithAllowedCallers(callers ...string) *App {
+	a.allowedCallers = callers
+	return a
+}
+
+// WithInput adds a custom input source to the App.
+func (a *App) WithInput(reader io.ReadCloser) *App {
+	a.input = reader
+	return a
+}
+
+// WithOutput adds a custom output sink to the App.
+func (a *App) WithOutput(writer io.WriteCloser) *App {
+	a.output = writer
+	return a
+}
+
+// AddHandler adds a new handler to the App for a given method.
+// There is at most one handler per method.
+func (a *App) AddHandler(method string, handler RequestHandler) {
+	a.handlers[method] = handler
+}
+
+// logError is a convenience method that logs an error and returns the same error.
+func (a *App) logError(err error) error {
+	if a.log != nil && err != nil {
+		a.log.Error(err.Error())
+	}
+	return err
+}
+
+// Run executes the helper application process.
+func (a *App) Run() error {
+	parentName, err := a.process.ParentProcessName()
+	if err != nil {
+		return a.logError(err)
+	}
+
+	if !a.isCallerPermitted(parentName) {
+		return a.logError(errors.Errorf("%s (version %s) may only be invoked by: %s",
+			a.Name(), build.DaosVersion, strings.Join(a.allowedCallers, ", ")))
+	}
+
+	// set up the r/w pipe from the parent process
+	conn := pbin.NewStdioConn(a.Name(), parentName, a.input, a.output)
+
+	if err = a.checkPrivileges(); err != nil {
+		resp := NewResponseWithError(a.logError(err))
+		sendErr := a.writeResponse(resp, conn)
+		if sendErr != nil {
+			return a.logError(sendErr)
+		}
+
+		return err
+	}
+
+	req, err := a.readRequest(conn)
+	if err != nil {
+		return a.logError(err)
+	}
+
+	resp := a.handleRequest(req)
+
+	err = a.writeResponse(resp, conn)
+	if err != nil {
+		return a.logError(err)
+	}
+
+	if resp.Error != nil {
+		return resp.Error
+	}
+	return nil
+}
+
+func (a *App) isCallerPermitted(callerName string) bool {
+	if len(a.allowedCallers) == 0 {
+		return true
+	}
+	for _, name := range a.allowedCallers {
+		if callerName == name {
+			return true
+		}
+	}
+	return false
+}
+
+func (a *App) checkPrivileges() error {
+	if !a.process.IsPrivileged() {
+		return pbin.PrivilegedHelperNotPrivileged(a.Name())
+	}
+
+	// hack for stuff that doesn't use geteuid() (e.g. ipmctl)
```

Comment:
(style)  'doesn'' may be misspelled - perhaps 'doesn't'?

Reply:
False positive

Reply:
@ashleypittman: Is this something we can tweak?

Reply:
Yes, I'm working on it.